pkg/commands/apply/apply_run.go

```diff
@@ -0,0 +1,500 @@
+// a Guardian apply using the Terraform CLI.
+func (c *ApplyRunCommand) terraformApply(ctx context.Context) (*RunResult, error) {
+	var stdout, stderr strings.Builder
+	multiStdout := io.MultiWriter(c.Stdout(), &stdout)
+	multiStderr := io.MultiWriter(c.Stderr(), &stderr)
+
+	c.Outf("Running Terraform commands")
+
+	lockfileMode := "none"
+	if !c.flagAllowLockfileChanges {
+		lockfileMode = "readonly"
+	}
+
+	if err := c.withActionsOutGroup("Initializing Terraform", func() error {
+		_, err := c.terraformClient.Init(ctx, c.Stdout(), multiStderr, &terraform.InitOptions{
+			Input:       util.Ptr(false),
+			NoColor:     util.Ptr(true),
+			Lockfile:    util.Ptr(lockfileMode),
+			LockTimeout: util.Ptr(c.flagLockTimeout.String()),
+		})
+		return err //nolint:wrapcheck // Want passthrough
+	}); err != nil {
+		return &RunResult{commentDetails: stderr.String()}, fmt.Errorf("failed to initialize: %w", err)
+	}
+
+	stderr.Reset()
+
+	if err := c.withActionsOutGroup("Validating Terraform", func() error {
+		_, err := c.terraformClient.Validate(ctx, c.Stdout(), multiStderr, &terraform.ValidateOptions{NoColor: util.Ptr(true)})
+		return err //nolint:wrapcheck // Want passthrough
+	}); err != nil {
+		return &RunResult{commentDetails: stderr.String()}, fmt.Errorf("failed to validate: %w", err)
+	}
+
+	stderr.Reset()
+
+	if err := c.withActionsOutGroup("Applying Terraform", func() error {
+		_, err := c.terraformClient.Apply(ctx, multiStdout, multiStderr, &terraform.ApplyOptions{
+			File:            util.Ptr(c.planFilename),
+			CompactWarnings: util.Ptr(true),
+			Input:           util.Ptr(false),
+			NoColor:         util.Ptr(true),
+			LockTimeout:     util.Ptr(c.flagLockTimeout.String()),
+		})
+
+		return err //nolint:wrapcheck // Want passthrough
+	}); err != nil {
+		return &RunResult{commentDetails: stderr.String()}, fmt.Errorf("failed to apply: %w", err)
+	}
+
+	stderr.Reset()
+
+	c.Outf("Formatting output")
+	githubOutput := terraform.FormatOutputForGitHubDiff(stdout.String())
+
+	return &RunResult{commentDetails: githubOutput}, nil
+}
+
+// withActionsOutGroup runs a function and ensures it is wrapped in GitHub actions
+// grouping syntax. If this is not in an action, output is printed without grouping syntax.
+func (c *ApplyRunCommand) withActionsOutGroup(msg string, fn func() error) error {
+	if c.GitHubFlags.FlagIsGitHubActions {
+		c.actions.Group(msg)
+		defer c.actions.EndGroup()
+	} else {
+		c.Outf(msg)
+	}
+
+	return fn()
+}
+
+// downloadGuardianPlan downloads the Guardian plan binary from the configured Guardian storage bucket
+// and returns the plan data and plan exit code.
+func (c *ApplyRunCommand) downloadGuardianPlan(ctx context.Context, path string) (planData []byte, planExitCode string, outErr error) {
+	c.Outf("Downloading Guardian plan file")
+
+	metadata, err := c.storageClient.ObjectMetadata(ctx, c.flagBucketName, path)
+	if err != nil {
+		outErr = fmt.Errorf("failed to get plan object metadata: %w", err)
```

Comment:
I don't think this actually got resolved. Unresolving.

Reply:
There's a defer below where he's trying to capture the error returned from calling `Close`.

Reply:
Yeah I see that, and I agree with the goal of what this is accomplishing. It's just that the way we're going about this can be simplified and made more conventional.

If we do `return nil, "", err`, then the value of err will be copied to outErr (since outErr is the declared name of the returned error at line 453). There's no need to manually assign to outErr. So I think we should stick with convention and just return the error.